extension/oauth2clientauthextension/clientcredentialsconfig.go

```diff
@@ -0,0 +1,105 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package oauth2clientauthextension // import "github.com/open-telemetry/opentelemetry-collector-contrib/extension/oauth2clientauthextension"
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+
+	"go.uber.org/multierr"
+	"golang.org/x/oauth2"
+	"golang.org/x/oauth2/clientcredentials"
+)
+
+// clientCredentialsConfig is a clientcredentials.Config wrapper to allow
+// values read from files in the ClientID and ClientSecret fields.
+//
+// Values from files can be retrieved by populating the ClientIDFile or
+// the ClientSecretFile fields with the path to the file.
+//
+// Priority: File > Raw value
+//
+// Example - Retrieve secret from file:
+//
+//	cfg := clientCredentialsConfig{
+//		Config: clientcredentials.Config{
+//			ClientID:     "clientId",
+//			...
+//		},
+//		ClientSecretFile: "/path/to/client/secret",
+//	}
+type clientCredentialsConfig struct {
+	clientcredentials.Config
+
+	ClientIDFile     string
+	ClientSecretFile string
+}
+
+type clientCredentialsTokenSource struct {
+	ctx    context.Context
+	config *clientCredentialsConfig
+}
+
+// clientCredentialsTokenSource implements TokenSource
+var _ oauth2.TokenSource = (*clientCredentialsTokenSource)(nil)
+
+func readCredentialsFile(path string) (string, error) {
+	f, err := os.ReadFile(path)
+	if err != nil {
+		return "", fmt.Errorf("failed to read credentials file %q: %w", path, err)
+	}
+
+	credential := strings.TrimSpace(string(f))
+	if credential == "" {
+		return "", fmt.Errorf("empty credentials file %q", path)
+	}
+	return credential, nil
+}
+
+func getActualValue(value, filepath string) (string, error) {
+	actualValue := value
+	var err error
+
+	if len(filepath) > 0 {
+		actualValue, err = readCredentialsFile(filepath)
```

Comment:
Do you need to create a new var? Perhaps just return the results of the readCredentialsFile if the filepath's len is > 0, otherwise return value, nil ?

```suggestion
		return readCredentialsFile(filepath)
```

Reply:
At this point we don't create a new var, we override the `actualValue` one (that's why we also define `err` outside the block). This doesn't look optimal at this state (and I'm good with changing it) but it makes things more clear in my opinion when adding more ways of retrieving the "actual value" (as mentioned in the issue, get by executing a command).

I can push the commits with the command part so that you can take a look and decide how you want me to change it.

Reply:
@jpkrohling this is the relevant commit adding the command case: https://github.com/elikatsis/opentelemetry-collector-contrib/commit/e453109b958265f58ca71b37701a5b7797469b18

Reply:
> as mentioned in the issue, get by executing a command

I missed that, and this specific example is unlikely to get accepted: we probably do not want the collector executing commands.

> At this point we don't create a new var, we override the actualValue one

I mean something like this:

```golang
func getActualValue(value, cmd, filepath string) (string, error) {
	if len(filepath) > 0 {
		return readCredentialsFile(filepath)
	}

	if len(cmd) > 0 {
		return readCommandOutput(cmd)
	}

	return value, nil
}
```

Reply:
By the way, this is a minor improvement, optional for the approval of this PR.

Another take would be this: given that this is in the hot path and the decision will be the same during the whole lifecycle of this component, it could even be better to just use a func as a parameter to this type. Like this:

```golang
func NewClientCredentialsConfig(cfg *clientcredentials.Config, clientIDFile, clientSecretFile string) (*clientCredentialsConfig, error) {
	c := &clientCredentialsConfig{
		Config: *cfg,
	}

	if len(clientIDFile) > 0 {
		c.clientIDSource = func() (string, error) {
			return readCredentialsFile(clientIDFile)
		}
	} else {
		c.clientIDSource = func() (string, error) {
			return c.ClientID, nil
		}
	}

	if len(clientSecretFile) > 0 {
		c.clientSecretSource = func() (string, error) {
			return readCredentialsFile(clientSecretFile)
		}
	} else {
		c.clientSecretSource = func() (string, error) {
			return c.ClientSecret, nil
		}
	}

	return c, nil
}
```

Your createConfig would then call `clientID, err := c.clientIDSource()` and `clientSecret, err := c.clientSecretSource()` accordingly.